### PR TITLE
7303 dynamic metaslab selection

### PR DESCRIPTION
**Pull request by grwilson**
Reviewed by: Chris Siden christopher.siden@delphix.com
Reviewed by: Matthew Ahrens mahrens@delphix.com
Reviewed by: Alex Reece alex@delphix.com
Reviewed by: Prakash Surya prakash.surya@delphix.com
Reviewed by: Paul Dagnelie paul.dagnelie@delphix.com
Reviewed by: Dan Kimmel dan.kimmel@delphix.com
Reviewed by: Pavel Zakharov pavel.zakharov@delphix.com

This change introduces a new weighting algorithm to improve metaslab selection. The new weighting algorithm relies on the SPACEMAP_HISTOGRAM feature. As a result, the metaslab weight now encodes the type of weighting algorithm used (size-based vs segment-based).

This also introduce a new allocation tracing facility and two new dcmds to help debug allocation problems. Each zio now contains a zio_alloc_list_t structure that is populated as the zio goes through the allocations stage. Here's an example of how to use the tracing facility:

<pre>
> c5ec000::print zio_t io_alloc_list | ::walk list | ::metaslab_trace
  MSID    DVA    ASIZE      WEIGHT             RESULT               VDEV
     -      0      400           0    NOT_ALLOCATABLE           ztest.0a
     -      0      400           0    NOT_ALLOCATABLE           ztest.0a
     -      0      400           0             ENOSPC           ztest.0a
     -      0      200           0    NOT_ALLOCATABLE           ztest.0a
     -      0      200           0    NOT_ALLOCATABLE           ztest.0a
     -      0      200           0             ENOSPC           ztest.0a
     1      0      400      1 x 8M            17b1a00           ztest.0a

> 1ff2400::print zio_t io_alloc_list | ::walk list | ::metaslab_trace
  MSID    DVA    ASIZE      WEIGHT             RESULT               VDEV
     -      0      200           0    NOT_ALLOCATABLE           mirror-2
     -      0      200           0    NOT_ALLOCATABLE           mirror-0
     1      0      200      1 x 4M            112ae00           mirror-1
     -      1      200           0    NOT_ALLOCATABLE           mirror-2
     -      1      200           0    NOT_ALLOCATABLE           mirror-0
     1      1      200      1 x 4M            112b000           mirror-1
     -      2      200           0    NOT_ALLOCATABLE           mirror-2</pre>

If the metaslab is using segment-based weighting then the WEIGHT column will display the number of segments available in the bucket where the allocation attempt was made.

Upstreams:
33015 Dynamic block allocator should be aware of dirty data
36458 remove ms_ops
35608 dynamic metaslab selection
39079 ztest should force more normal allocations
DLPX-33311 metaslab_group_alloc_normal is O(n^2), causes assertion failure
DLPX-33312 metaslab_alloc: trying all possible distances causes trace overflow
DLPX-35383 metaslab load/unload every txg causes 20x performance degradation
DLPX-34472 metaslab trace records can be leaked
DLPX-36342 NULL pointer dereference when activating a partially added metaslab
DLPX-40599 crash in zio_nowait due to metaslab_preload while spa_unload
DLPX-46725 metaslab_trace has unused heading

### CODE REVIEW

**Comment by ahrens** [comment]
@zettabot go

File: usr/src/uts/common/fs/zfs/sys/metaslab.h [hunk 1]
```diff
@@ -20,7 +20,7 @@
  */
 /*
  * Copyright (c) 2005, 2010, Oracle and/or its affiliates. All rights reserved.
- * Copyright (c) 2011, 2015 by Delphix. All rights reserved.
+ * Copyright (c) 2011, 2016 by Delphix. All rights reserved.
```
Comment (don-brady) [inline usr/src/uts/common/fs/zfs/sys/metaslab.h]: Copyright bummped here but not in metaslab.c or metaslab_impl.h

File: usr/src/uts/common/fs/zfs/metaslab.c [hunk 2]
```diff
@@ -2143,6 +2473,113 @@ metaslab_distance(metaslab_t *msp, dva_t *dva)
 
 /*
  * ==========================================================================
+ * Metaslab allocation tracing facility
```
Comment (don-brady) [inline usr/src/uts/common/fs/zfs/metaslab.c]: FYI.  For zfsonlinux we conditionally removed the tracing since no imediate way to observe it.

**Comment by ahrens** [comment]
@zettabot go

**Pull request closed without merging**